File: integrations/kotlin/src/jsMain/kotlin/angular/ts/ProgrammaticView.kt
package angular.ts

import angular.ts.generated.Angular as RawAngular
import angular.ts.generated.ProgrammaticViewContext as RawProgrammaticViewContext
import angular.ts.generated.ProgrammaticViewTags as RawProgrammaticViewTags
import angular.ts.generated.Scope as RawScope
import org.w3c.dom.Element
import org.w3c.dom.HTMLElement

public class ProgrammaticViewContext<TController, TRequired> internal constructor(
    internal val raw: RawProgrammaticViewContext<TController, TRequired, RawScope, HTMLElement>,
) {
    public val controller: TController?
        get() = raw.controller.unsafeCast<TController?>()

    public val required: TRequired?
        get() = raw.required.unsafeCast<TRequired?>()

    public val scope: Scope<Any>
        get() = Scope(raw.scope.unsafeCast<RawScope>())

    public val host: HTMLElement
        get() = raw.asDynamic().host.unsafeCast<HTMLElement>()

    public val transclude: dynamic
        get() = raw.asDynamic().transclude

    public fun onDestroy(cleanup: () -> Unit): () -> Unit =
        raw.onDestroy(cleanup)
}

public typealias ProgrammaticView<TController, TRequired> =
    (ProgrammaticViewContext<TController, TRequired>) -> dynamic

internal fun <TController, TRequired> ProgrammaticView<TController, TRequired>.toJs(): dynamic =
    { context: dynamic ->
        this(
            ProgrammaticViewContext(
                context.unsafeCast<
                    RawProgrammaticViewContext<TController, TRequired, RawScope, HTMLElement>,
                >(),
            ),
        )
    }

public class ProgrammaticTags internal constructor(
    internal val raw: RawProgrammaticViewTags,
) {
    public fun namespace(namespaceUri: String): ProgrammaticTags =
        ProgrammaticTags(raw(namespaceUri).unsafeCast<RawProgrammaticViewTags>())

    public fun tag(
        name: String,
        properties: Map<String, Any?> = emptyMap(),
        vararg children: Any?,
    ): Element {
        val arguments = arrayOf(properties.toJsRecord(), *children)

        return callJsFunction(raw.asDynamic()[name], null, arguments).unsafeCast<Element>()
    }
}

public class ProgrammaticViewApi internal constructor(
    internal val raw: RawAngular,
) {
    public val tags: ProgrammaticTags
        get() = ProgrammaticTags(raw.view.tags)

    public fun event(
        listener: (dynamic) -> Unit,
        options: dynamic = undefined,
    ): dynamic = callJsFunction(raw.view.event, null, arrayOf(listener, options))

    public fun attrs(values: Map<String, Any?>): dynamic =
        callJsFunction(raw.view.attrs, null, arrayOf(values.toJsRecord()))

    public fun props(values: Map<String, Any?>): dynamic =
        callJsFunction(raw.view.props, null, arrayOf(values.toJsRecord()))

    public fun <T> each(
        read: () -> Array<T>?,
        key: (T) -> Any,
        render: (() -> T) -> Any?,
    ): dynamic = callJsFunction(raw.view.each, null, arrayOf(read, key, render))

    public fun tag(
        name: String,
        properties: Map<String, Any?> = emptyMap(),
        vararg children: Any?,
    ): Element = callJsFunction(
        raw.view.tag,
        null,
        arrayOf(name, properties.toJsRecord(), *children),
    ).unsafeCast<Element>()

    public fun tagNS(
        namespaceUri: String,
        name: String,
        properties: Map<String, Any?> = emptyMap(),
        vararg children: Any?,
    ): Element = callJsFunction(
        raw.view.tagNS,
        null,
        arrayOf(namespaceUri, name, properties.toJsRecord(), *children),
    ).unsafeCast<Element>()
}

File: integrations/scala/src/main/scala/angular/ts/ProgrammaticView.scala
package angular.ts

import org.scalajs.dom.HTMLElement
import scala.scalajs.js
import scala.scalajs.js.JSConverters.*

@js.native
trait ProgrammaticViewContext[+Controller, +Required] extends js.Object:
  val controller: js.UndefOr[Controller] = js.native

  val required: js.UndefOr[Required] = js.native

  val scope: Scope = js.native

  val host: HTMLElement = js.native

  val transclude: js.Function = js.native

  val onDestroy: js.Function1[js.Function0[Unit], js.Function0[Unit]] = js.native

type ProgrammaticView[Controller, Required] =
  js.Function1[ProgrammaticViewContext[Controller, Required], js.Any]

final class ProgrammaticTags private[ts] (private val raw: js.Dynamic):
  def namespace(namespaceUri: String): ProgrammaticTags =
    ProgrammaticTags(
      RuntimeReflect
        .invoke(raw.asInstanceOf[js.Function], js.undefined, js.Array(namespaceUri))
        .asInstanceOf[js.Dynamic],
    )

  def tag(
      name: String,
      properties: Map[String, js.Any] = Map.empty,
      children: js.Any*,
  ): org.scalajs.dom.Element =
    val arguments = js.Array[js.Any](properties.toJSDictionary)
    arguments.push(children*)

    RuntimeReflect
      .invoke(raw.selectDynamic(name).asInstanceOf[js.Function], js.undefined, arguments)
      .asInstanceOf[org.scalajs.dom.Element]

final class ProgrammaticViewApi private[ts] (private val raw: js.Dynamic):
  def event(
      listener: js.Function1[org.scalajs.dom.Event, Unit],
      options: js.Any = js.undefined,
  ): js.Function1[org.scalajs.dom.Event, Unit] =
    RuntimeReflect
      .invoke(raw.event.asInstanceOf[js.Function], js.undefined, js.Array(listener, options))
      .asInstanceOf[js.Function1[org.scalajs.dom.Event, Unit]]

  def attrs(values: Map[String, js.Any]): js.Dynamic =
    RuntimeReflect
      .invoke(raw.attrs.asInstanceOf[js.Function], js.undefined, js.Array(values.toJSDictionary))
      .asInstanceOf[js.Dynamic]

  def props(values: Map[String, js.Any]): js.Dynamic =
    RuntimeReflect
      .invoke(raw.props.asInstanceOf[js.Function], js.undefined, js.Array(values.toJSDictionary))
      .asInstanceOf[js.Dynamic]

  def each[T](
      read: js.Function0[js.Array[T]],
      key: js.Function1[T, js.Any],
      render: js.Function1[js.Function0[T], js.Any],
  ): js.Function0[js.Any] =
    RuntimeReflect
      .invoke(raw.each.asInstanceOf[js.Function], js.undefined, js.Array(read, key, render))
      .asInstanceOf[js.Function0[js.Any]]

  def tag(
      name: String,
      properties: Map[String, js.Any] = Map.empty,
      children: js.Any*,
  ): org.scalajs.dom.Element =
    val arguments = js.Array[js.Any](name, properties.toJSDictionary)
    arguments.push(children*)
    RuntimeReflect
      .invoke(raw.tag.asInstanceOf[js.Function], js.undefined, arguments)
      .asInstanceOf[org.scalajs.dom.Element]

  def tagNS(
      namespaceUri: String,
      name: String,
      properties: Map[String, js.Any] = Map.empty,
      children: js.Any*,
  ): org.scalajs.dom.Element =
    val arguments = js.Array[js.Any](namespaceUri, name, properties.toJSDictionary)
    arguments.push(children*)
    RuntimeReflect
      .invoke(raw.tagNS.asInstanceOf[js.Function], js.undefined, arguments)
      .asInstanceOf[org.scalajs.dom.Element]
